File: Classes/View Controllers/Tabs/Home/HomeAlbumViewController.m
//
//  HomeAlbumViewController.m
//  iSub
//
//  Created by Ben Baron on 2/28/10.
//  Copyright 2010 Ben Baron. All rights reserved.
//

#import "HomeAlbumViewController.h"
#import "ServerListViewController.h"
#import "Defines.h"
#import "SavedSettings.h"
#import "MusicSingleton.h"
#import "EX2Kit.h"
#import "Swift.h"

@interface HomeAlbumViewController() <APILoaderDelegate>
@end

@implementation HomeAlbumViewController

- (instancetype)initWithNibName:(NSString *)n bundle:(NSBundle *)b; {
    if (self = [super initWithNibName:n bundle:b]) {
		_isMoreAlbums = YES;
    }
    return self;
}

- (void)viewDidLoad {
	[super viewDidLoad];
	    
    self.tableView.rowHeight = Defines.rowHeight;
    [self.tableView registerClass:UniversalTableViewCell.class forCellReuseIdentifier:UniversalTableViewCell.reuseId];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self addShowPlayerButton];
}

- (void)loadMoreResults {
	if (self.isLoading) return;
	
	self.isLoading = YES;
	self.offset += 20;
    
    self.loader = [[QuickAlbumsLoader alloc] initWithDelegate:self];
    self.loader.modifier = self.modifier;
    self.loader.offset = self.offset;
    [self.loader startLoad];
}

- (void)loadingFinished:(APILoader *)loader {
    if (self.loader.folderAlbums.count == 0) {
        // There are no more songs
        self.isMoreAlbums = NO;
    } else {
        // Add the new results to the list of songs
        [self.folderAlbums addObjectsFromArray:self.loader.folderAlbums];
    }
    
    // Reload the table
    [self.tableView reloadData];
    self.isLoading = NO;
    
    self.loader = nil;
}

- (void)loadingFailed:(APILoader *)theLoader error:(NSError *)error {
    self.loader = nil;
	self.isLoading = NO;
	    
    if (settingsS.isPopupsEnabled) {
        NSString *message = [NSString stringWithFormat:@"There was an error performing the search.\n\nError:%@", error.localizedDescription];
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error" message:message preferredStyle:UIAlertControllerStyleAlert];
        [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
        [self presentViewController:alert animated:YES completion:nil];
    }
}

#pragma mark Table view methods

// Customize the number of rows in the table view.
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section  {
	return self.folderAlbums.count + 1;
}

// Customize the appearance of table view cells.
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath  {
	if (indexPath.row < self.folderAlbums.count) {
        UniversalTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:UniversalTableViewCell.reuseId];
        [cell showCached:NO number:NO art:YES secondary:YES duration:NO];
        [cell updateWithModel:[self.folderAlbums objectAtIndexSafe:indexPath.row]];
        return cell;
	} else if (indexPath.row == self.folderAlbums.count) {
		// This is the last cell and there could be more results, load the next 20 songs;
		UITableViewCell *cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"HomeAlbumLoadCell"];
        cell.backgroundColor = [UIColor colorNamed:@"isubBackgroundColor"];
		if (self.isMoreAlbums) {
			cell.textLabel.text = @"Loading more results...";
            UIActivityIndicatorView *indicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleMedium];
			indicator.center = CGPointMake(300, 30);
			[cell addSubview:indicator];
			[indicator startAnimating];
            
			[self loadMoreResults];
		} else {
			cell.textLabel.text = @"No more results";
		}
		
		return cell;
	}
	
	// In case somehow no cell is created, return an empty cell
	return [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier: @"EmptyCell"];
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
	if (!indexPath) return;
	
	if (indexPath.row != self.folderAlbums.count) {
		ISMSFolderAlbum *folderAlbum = [self.folderAlbums objectAtIndexSafe:indexPath.row];
		FolderAlbumViewController *controller = [[FolderAlbumViewController alloc] initWithFolderAlbum:folderAlbum];
		[self pushViewControllerCustom:controller];
	} else {
		[self.tableView deselectRowAtIndexPath:indexPath animated:NO];
	}
}

- (UISwipeActionsConfiguration *)tableView:(UITableView *)tableView trailingSwipeActionsConfigurationForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row != self.folderAlbums.count) {
        return [SwipeAction downloadAndQueueConfigWithModel:[self.folderAlbums objectAtIndexSafe:indexPath.row]];
    }
    return nil;
}

@end

File: Classes/View Controllers/Tabs/Home/HomeAlbumViewController.swift
//
//  HomeAlbumViewController.swift
//  iSub
//
//  Created by Benjamin Baron on 1/19/21.
//  Copyright © 2021 Ben Baron. All rights reserved.
//

import UIKit
import Resolver

final class HomeAlbumViewController: UIViewController {
    @Injected private var settings: SavedSettings
    
    var serverId: Int { (Resolver.resolve() as SavedSettings).currentServerId }
    
    let modifier: QuickAlbumsModifier
    var folderAlbums: [FolderAlbum]
    
    private let tableView = UITableView()
    
    private var isMoreAlbums: Bool
    private var isLoading = false
    private var offset = 0
    private var loaderTask: Task<Void, Never>?
    
    init(modifier: QuickAlbumsModifier, folderAlbums: [FolderAlbum], title: String) {
        self.modifier = modifier
        self.folderAlbums = folderAlbums
        self.isMoreAlbums = folderAlbums.count >= AsyncSearchLoader.searchItemCount
        super.init(nibName: nil, bundle: nil)
        self.title = title
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        view.backgroundColor = Colors.background
        setupDefaultTableView(tableView)
    }
    
    deinit {
        loaderTask?.cancel()
    }
    
    private func loadMoreResults() {
        guard !isLoading else { return }
        
        isLoading = true
        offset += 20
        
        loaderTask = Task {
            do {
                defer {
                    tableView.reloadData()
                    isLoading = false
                }
                
                let responseFolderAlbums = try await AsyncQuickAlbumsLoader(serverId: serverId, modifier: modifier, offset: offset).load()
                if responseFolderAlbums.count == 0 {
                    isMoreAlbums = false
                } else {
                    folderAlbums.append(contentsOf: responseFolderAlbums)
                }
            } catch {
                if self.settings.isPopupsEnabled {
                    let message = "There was an error performing the search.\n\nError: \(error.localizedDescription)"
                    let alert = UIAlertController(title: "Error", message: message, preferredStyle: .alert)
                    alert.addOKAction()
                    self.present(alert, animated: true, completion: nil)
                }
            }
        }
    }
}

extension HomeAlbumViewController: UITableViewConfiguration {
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return folderAlbums.count + 1
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if indexPath.row < folderAlbums.count {
            let cell = tableView.dequeueUniversalCell()
            cell.show(downloaded: false, number: false, art: true, secondary: true, duration: false)
            cell.update(model: folderAlbums[indexPath.row])
            return cell
        } else {
            // This is the last cell and there could be more results, load the next 20 songs;
            let cell = UITableViewCell(style: .default, reuseIdentifier: "HomeAlbumLoadCell")
            cell.backgroundColor = .systemBackground
            if isMoreAlbums {
                cell.textLabel?.text = "Loading more results...";
                let indicator = UIActivityIndicatorView(style: .medium)
                indicator.center = CGPoint(x: 300, y: tableView.rowHeight / 2.0)
                indicator.autoresizingMask = .flexibleLeftMargin
                cell.addSubview(indicator)
                indicator.startAnimating()
                loadMoreResults()
            } else {
                cell.textLabel?.text = "No more results"
            }
            return cell
        }
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if indexPath.row < folderAlbums.count {
            let controller = FolderAlbumViewController(folderAlbum: folderAlbums[indexPath.row])
            pushViewControllerCustom(controller)
        } else {
            tableView.deselectRow(at: indexPath, animated: false)
        }
    }
    
    func tableView(_ tableView: UITableView, trailingSwipeActionsConfigurationForRowAt indexPath: IndexPath) -> UISwipeActionsConfiguration? {
        if indexPath.row < folderAlbums.count {
            return SwipeAction.downloadAndQueueConfig(model: folderAlbums[indexPath.row])
        }
        return nil
    }
}
